arangod/RocksDBEngine/RocksDBVectorIndex.cpp

```diff
@@ -0,0 +1,470 @@
+    return _it->Valid() && _it->key().starts_with(_rocksdbKey.string());
+  }
+
+  void next() override { _it->Next(); }
+
+  std::pair<faiss::idx_t, uint8_t const*> get_id_and_codes() override {
+    auto const docId = RocksDBKey::indexDocumentId(_it->key());
+    TRI_ASSERT(_codeSize == _it->value().size());
+    auto const* value = reinterpret_cast<uint8_t const*>(_it->value().data());
+    return {static_cast<faiss::idx_t>(docId.id()), value};
+  }
+
+ private:
+  RocksDBKey _rocksdbKey;
+  arangodb::RocksDBVectorIndex* _index = nullptr;
+
+  std::unique_ptr<rocksdb::Iterator> _it;
+  std::size_t _listNumber;
+  std::size_t _codeSize;
+};
+
+struct RocksDBInvertedLists : faiss::InvertedLists {
+  RocksDBInvertedLists(RocksDBVectorIndex* index, LogicalCollection* collection,
+                       transaction::Methods* trx,
+                       RocksDBMethods* rocksDBMethods,
+                       rocksdb::ColumnFamilyHandle* cf, std::size_t nlist,
+                       size_t codeSize)
+      : InvertedLists(nlist, codeSize),
+        _index(index),
+        _collection(collection),
+        _trx(trx),
+        _rocksDBMethods(rocksDBMethods),
+        _cf(cf) {
+    use_iterator = true;
+    assert(status.ok());
+  }
+
+  std::size_t list_size(std::size_t /*listNumber*/) const override {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_NOT_IMPLEMENTED,
+                                   "faiss list_size not supported");
+  }
+
+  std::uint8_t const* get_codes(std::size_t /*listNumber*/) const override {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_NOT_IMPLEMENTED,
+                                   "faiss get_codes not supported");
+  }
+
+  faiss::idx_t const* get_ids(std::size_t /*listNumber*/) const override {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_NOT_IMPLEMENTED,
+                                   "faiss get_ids not supported");
+  }
+
+  size_t add_entries(std::size_t listNumber, std::size_t nEntry,
+                     faiss::idx_t const* ids,
+                     std::uint8_t const* code) override {
+    for (std::size_t i = 0; i < nEntry; i++) {
+      RocksDBKey rocksdbKey;
+      // Can we get away with saving LocalDocumentID in key, maybe we need to
+      // the use IdSelector in options during search
+      auto const docId = LocalDocumentId(*(ids + i));
+      rocksdbKey.constructVectorIndexValue(_index->objectId(), listNumber,
+                                           docId);
+
+      auto const value = RocksDBValue::VectorIndexValue(
+          reinterpret_cast<const char*>(code + i * code_size), code_size);
+      auto const status =
+          _rocksDBMethods->PutUntracked(_cf, rocksdbKey, value.string());
+
+      if (!status.ok()) {
+        // Here we need to throw since there is no way to return the status
+        auto const res = rocksutils::convertStatus(status);
+        THROW_ARANGO_EXCEPTION_MESSAGE(res.errorNumber(), res.errorMessage());
+      }
+    }
+    return 0;
```

Comment:
If I understand the faiss code correctly, i.e. quoting from the comment above `faiss::InvertedIndex`
```
/** Table of inverted lists
 * multithreading rules:
 * - concurrent read accesses are allowed
 * - concurrent update accesses are allowed
 * - for resize and add_entries, only concurrent access to different lists
 *   are allowed
 */
```
but we use the same rocksdb transaction object and so we have concurrent access to the rocksdb methods. I don't think that these are thread-safe to use.

Reply:
The reason I think `RocksDBMethods` used here are thread-safe is because they are used in the same way in other indexes as well. In `RocksdBVectorIndex::insert` we pass the RocksDBMethods to the InvertedList once, and that methods inserts a single document.